File: SampSharpGameMode/Story/Actions/SitDown.cs
using SampSharp.GameMode;
using SampSharp.SyntheticGameMode.Data;
using SampSharp.SyntheticGameMode.Extensions;
using SampSharpGameMode.Extensions;
using SyntheticVideo2language.Story.Api;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace SampSharp.SyntheticGameMode.Story.Actions
{
    public class SitDown : StoryActionBase
    {
        public enum eHow
        {
            atDesk,
            sofa
        }

        public override string Description { get => " sits down "; set { } }
        public override IStoryActor Performer { get; set; }
        public override IStoryItem TargetItem { get; set; }
        public eHow How { get; set; }

        public override async Task<bool> ApplyAsync(params object[] parameters)
        {
            SampStory.Instance.History.Add(this);

            await Task.Delay(100);
            var player = Performer as Player;
            var targetObject = (TargetItem as Objects.SampStoryObjectBase);
            SampStory.Instance.Logger.Log(player.Description + " " + this.Description + " on the " + TargetItem.Description, player);
            //TODO: move the player in the middle of the target object with the face in the opposite direction
            switch (this.How)
            {
                case eHow.atDesk:
                    player.ApplyAnimation("INT_OFFICE", "OFF_Sit_In", 4.1f, false, false, false, true, 5000, true);
                    break;
                case eHow.sofa:
                    //var bbox = targetObject.GetBoundingBox();
                    //player.SendClientMessage((bbox.Max.X - bbox.Min.X).ToString());
                    //player.SendClientMessage((bbox.Max.Y - bbox.Min.Y).ToString());
                    //bbox.ChangeOrigin(targetObject.Position, new Vector3(0, 0, targetObject.Rotation.Z));
                    //var across = Vector3.UnitY.Rotate(new Vector3(0, 0, targetObject.Rotation.Z));
                    //var along = Vector3.UnitX.Rotate(new Vector3(0, 0, targetObject.Rotation.Z));
                    //var centerTopMiddle = bbox.Min.ProjectOnPlane(along.MovePointAlong(player.Position, 10).Normalized());
                    //var centerTopLeft = centerTopMiddle + across.Normalized().Mult((bbox.Max.X - bbox.Min.X) / 2);
                    //Debug.DrawPoint(player, centerTopLeft, Debug.ePointType.Type2);
                    //player.Position = new Vector3(centerTopLeft.X, centerTopLeft.Y, player.Position.Z);
                    player.Position = player.GetHeadingVector().MovePointAlong(player.Position, -0.6f);
                    player.ApplyAnimation("INT_HOUSE", "LOU_In", 4.1f, false, false, false, true, 5000, true);
                    break;
                default:
                    break;
            }
            Thread.Sleep(5000);

            return await (NextLocation as Location).GetNextRandomValidAction().ApplyAsync();
        }

        public SitDown(eHow how) : base()
        {
            this.How = how;
        }
    }
}

File: story/Actions/SitDown.lua
SitDown = class(StoryActionBase, function(o, params)
    -- check mandatory options
    -- if type(params.performer) ~= "userdata" then
    --     error("SitDown: performer not given in the constructor")
    -- elseif type(params.targetItem) ~= "table" then
    --     error("SitDown: targetItem not given in the constructor")
    -- elseif type(params.nextLocation) ~= "table" then
    --     error("SitDown: nextLocation not given in the constructor")
    -- end
    params.description = " sits down"
    params.name = 'SitDown'

    StoryActionBase.init(o,params)
    o.how = params.how or SitDown.eHow.atDesk
    o.rotation= params.rotation or nil
end)

SitDown.eHow = {
    atDesk = 1,
    onSofa = 2
}

function SitDown:Apply()
    local story = GetStory(self.Performer)
    table.insert(story.History[self.Performer:getData('id')], self)

    if self.Performer:getData("currentRegionId") == story.CurrentEpisode.CurrentRegion.Id then
        story.Logger:Log(self.Description .. " on the " .. self.TargetItem.Description, self)
    end
    
    self.TargetItem.instance:setCollisionsEnabled(false)

    if not (self.rotation == nil) then
        self.Performer.rotation = self.rotation
    end

    local animationLib = "INT_OFFICE"
    local animationId = "OFF_Sit_In"
    local duration = 5000
    if self.how == SitDown.eHow.atDesk then
        animationLib = "INT_OFFICE"
        animationId = "OFF_Sit_In"
        duration = 4000
    elseif self.how == SitDown.eHow.onSofa then
        animationLib = "INT_HOUSE"
        animationId = "LOU_In"
        self.Performer.position = self.Performer.matrix.position - self.Performer.matrix.forward * 0.6
        duration = 5000
    end

    self.Performer:setAnimation(animationLib, animationId, -1, false, true, false, true)
    if DEBUG then
        outputConsole("SitDown:Apply")
    end

    OnGlobalActionFinished(duration, self.Performer:getData('id'), self.Performer:getData('storyId'))
end

function SitDown:GetDynamicString()
    local rotationStr = nil
    if self.rotation then
        rotationStr = 'Vector3('..self.rotation.x..', '..self.rotation.y..', '..self.rotation.z..')'
    end
    return 'return SitDown{how = '..self.how..', rotation = '..(rotationStr or 'nil')..'}'
end